### PR TITLE
[WIP] Implemented base VRR Support

### PR DESCRIPTION
**Pull request by KiritoDv**

### CODE REVIEW

File: src/core/Window.cpp [hunk 1]
```diff
@@ -297,6 +297,18 @@ uint32_t Window::GetCurrentRefreshRate() {
...
 
+bool Window::HasVRRSupport() {
+    return mWindowManagerApi->has_vrr_support();
+}
+
+bool Window::GetVSyncStatus() {
+    return mWindowManagerApi->get_vsync();
+}
+
+void Window::SetVSyncStatus(bool status) {
```
Comment (Kenix3) [inline src/core/Window.cpp]:
```suggestion
void Window::SetVerticalSync(bool vsyncToggle) {
```

File: src/core/Window.cpp [hunk 2]
```diff
@@ -297,6 +297,18 @@ uint32_t Window::GetCurrentRefreshRate() {
     return mRefreshRate;
 }
 
+bool Window::HasVRRSupport() {
+    return mWindowManagerApi->has_vrr_support();
+}
+
+bool Window::GetVSyncStatus() {
```
Comment (Kenix3) [inline src/core/Window.cpp]:
```suggestion
bool Window::GetVerticalSync() {
```

File: src/core/Window.h [hunk 3]
```diff
@@ -48,6 +48,9 @@ class Window {
     uint32_t GetCurrentWidth();
     uint32_t GetCurrentHeight();
     uint32_t GetCurrentRefreshRate();
+    bool HasVRRSupport();
+    bool GetVSyncStatus();
```
Comment (Kenix3) [inline src/core/Window.h]:
```suggestion
    bool GetVerticalSync();
```

File: src/core/Window.h [hunk 4]
```diff
@@ -48,6 +48,9 @@ class Window {
...
+    void SetVSyncStatus(bool status);
```
Comment (Kenix3) [inline src/core/Window.h]:
```suggestion
    void SetVerticalSync(bool status);
```

**Review by Kenix3 — CHANGES_REQUESTED** [review]

File: src/core/Window.cpp [hunk 5]
```diff
@@ -297,6 +297,18 @@ uint32_t Window::GetCurrentRefreshRate() {
...
+bool Window::HasVRRSupport() {
+    return mWindowManagerApi->has_vrr_support();
+}
+
+bool Window::GetVerticalSync() {
+    return mWindowManagerApi->get_vsync();
+}
+
+void Window::SetVerticalSync(bool vsyncToggle) {
+    mWindowManagerApi->set_vsync(status);
```
Comment (Kenix3) [inline src/core/Window.cpp]:
```suggestion
    mWindowManagerApi->set_vsync(vsync);
```

File: src/core/Window.h [hunk 6]
```diff
@@ -48,6 +48,9 @@ class Window {
     uint32_t GetCurrentWidth();
     uint32_t GetCurrentHeight();
     uint32_t GetCurrentRefreshRate();
+    bool HasVRRSupport();
+    bool GetVerticalSync();
+    void SetVerticalSync(bool status);
```
Comment (Kenix3) [inline src/core/Window.h]:
```suggestion
    void SetVerticalSync(bool vsync);
```

File: src/core/Window.cpp [hunk 7]
```diff
@@ -297,6 +297,18 @@ uint32_t Window::GetCurrentRefreshRate() {
...
 
+bool Window::HasVRRSupport() {
+    return mWindowManagerApi->has_vrr_support();
+}
+
+bool Window::GetVerticalSync() {
+    return mWindowManagerApi->get_vsync();
+}
+
+void Window::SetVerticalSync(bool vsyncToggle) {
```
Comment (Kenix3) [inline src/core/Window.cpp]:
```suggestion
void Window::SetVerticalSync(bool vsync) {
```

**Review by Kenix3 — CHANGES_REQUESTED** [review]

**Comment by briaguya0** [comment]
this was last tested and confirmed working with https://github.com/HarbourMasters/Shipwright/actions/runs/4714449533, which was commit https://github.com/Kenix3/libultraship/pull/168/commits/a52b47edd0bb54019e9e1a91cf63058cb417201e

this means everything here is untested: https://github.com/Kenix3/libultraship/compare/a52b47edd0bb54019e9e1a91cf63058cb417201e...KiritoDv:libultraship:fix/vrr_support

File: src/core/Window.cpp [hunk 8]
```diff
@@ -297,6 +297,18 @@ uint32_t Window::GetCurrentRefreshRate() {
     return mRefreshRate;
 }
 
+bool Window::HasVRRSupport() {
```
Comment (Kenix3) [inline src/core/Window.cpp]:
```suggestion
bool Window::CanDisableVerticalSync() {
```

File: src/core/Window.cpp [hunk 9]
```diff
@@ -297,6 +297,18 @@ uint32_t Window::GetCurrentRefreshRate() {
...
+    return mWindowManagerApi->has_vrr_support();
```
Comment (Kenix3) [inline src/core/Window.cpp]:
```suggestion
    return mWindowManagerApi->can_disable_vsync();
```

File: src/core/Window.h [hunk 10]
```diff
@@ -48,6 +48,9 @@ class Window {
     uint32_t GetCurrentWidth();
     uint32_t GetCurrentHeight();
     uint32_t GetCurrentRefreshRate();
+    bool HasVRRSupport();
```
Comment (Kenix3) [inline src/core/Window.h]:
```suggestion
    bool CanDisableVerticalSync();
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 11]
```diff
@@ -761,6 +805,18 @@ const char* gfx_dxgi_get_key_name(int scancode) {
     return text;
 }
 
+bool gfx_dxgi_has_vrr_support() {
```
Comment (Kenix3) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
bool gfx_dxgi_can_disable_vsync() {
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 12]
```diff
@@ -777,6 +833,9 @@ extern "C" struct GfxWindowManagerAPI gfx_dxgi_api = { gfx_dxgi_init,
                                                        gfx_dxgi_get_time,
                                                        gfx_dxgi_set_target_fps,
                                                        gfx_dxgi_set_maximum_frame_latency,
-                                                       gfx_dxgi_get_key_name };
+                                                       gfx_dxgi_get_key_name,
+                                                       gfx_dxgi_has_vrr_support,
```
Comment (Kenix3) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
                                                       gfx_dxgi_can_disable_vsync,
```

File: src/graphic/Fast3D/gfx_sdl2.cpp [hunk 13]
```diff
@@ -545,6 +549,18 @@ static const char* gfx_sdl_get_key_name(int scancode) {
     return SDL_GetScancodeName((SDL_Scancode)untranslate_scancode(scancode));
 }
 
+bool gfx_sdl_has_vrr_support() {
```
Comment (Kenix3) [inline src/graphic/Fast3D/gfx_sdl2.cpp]:
```suggestion
bool gfx_sdl_can_disable_vsync() {
```

File: src/graphic/Fast3D/gfx_sdl2.cpp [hunk 14]
```diff
@@ -561,6 +577,9 @@ struct GfxWindowManagerAPI gfx_sdl = { gfx_sdl_init,
                                        gfx_sdl_get_time,
                                        gfx_sdl_set_target_fps,
                                        gfx_sdl_set_maximum_frame_latency,
-                                       gfx_sdl_get_key_name };
+                                       gfx_sdl_get_key_name,
+                                       gfx_sdl_has_vrr_support,
```
Comment (Kenix3) [inline src/graphic/Fast3D/gfx_sdl2.cpp]:
```suggestion
                                       gfx_sdl_can_disable_vsync,
```

File: src/graphic/Fast3D/gfx_window_manager_api.h [hunk 15]
```diff
@@ -24,6 +24,9 @@ struct GfxWindowManagerAPI {
     void (*set_target_fps)(int fps);
     void (*set_maximum_frame_latency)(int latency);
     const char* (*get_key_name)(int scancode);
+    bool (*has_vrr_support)();
```
Comment (Kenix3) [inline src/graphic/Fast3D/gfx_window_manager_api.h]:
```suggestion
    bool (*can_disable_vsync)();
```

**Review by Kenix3 — CHANGES_REQUESTED** [review]

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 16]
```diff
@@ -76,6 +77,9 @@ static struct {
     uint32_t applied_maximum_frame_latency;
     HANDLE timer;
     bool use_timer;
+    BOOL allow_tearing;
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 17]
```diff
@@ -667,11 +700,19 @@ void gfx_dxgi_create_factory_and_device(bool debug, int d3d_version,
...
+
+        ComPtr<IDXGIFactory5> factory;
+        HRESULT hr = dxgi.factory.As(&factory);
+
+        if (SUCCEEDED(hr)) {
+            hr = factory->CheckFeatureSupport(DXGI_FEATURE_PRESENT_ALLOW_TEARING, &dxgi.allow_tearing,
+                                              sizeof(dxgi.allow_tearing));
         }
+
+        dxgi.tearing_support = SUCCEEDED(hr) && dxgi.allow_tearing;
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
        BOOL allowTearing = FALSE;
        if (SUCCEEDED(hr)) {
            hr = factory->CheckFeatureSupport(DXGI_FEATURE_PRESENT_ALLOW_TEARING, &allowTearing,
                                              sizeof(allowTearing));
        }

        dxgi.tearing_support = SUCCEEDED(hr) && allowTearing;
```

**Review by briaguya0 — COMMENTED** [review]
i'd like these changes please

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 18]
```diff
@@ -76,6 +77,8 @@ static struct {
     uint32_t applied_maximum_frame_latency;
     HANDLE timer;
     bool use_timer;
+    bool tearing_support;
+    bool force_disable_vsync = true;
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
    bool vsync_enabled = true;
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 19]
```diff
@@ -422,150 +425,158 @@ static uint64_t qpc_to_100ns(uint64_t qpc) {
...
-            LARGE_INTEGER t0;
-            QueryPerformanceCounter(&t0);
-            // printf("Get frame stats: %llu\n", (unsigned long long)(t0.QuadPart - dxgi.qpc_init));
-        }
-        // printf("stats: %u %u %u %u %u %.6f\n", dxgi.pending_frame_stats.rbegin()->first,
-        // dxgi.pending_frame_stats.rbegin()->second, stats.PresentCount, stats.PresentRefreshCount,
-        // stats.SyncRefreshCount, (double)(stats.SyncQPCTime.QuadPart - dxgi.qpc_init) / dxgi.qpc_freq);
-        if (dxgi.frame_stats.empty() || dxgi.frame_stats.rbegin()->second.PresentCount != stats.PresentCount) {
-            dxgi.frame_stats.insert(std::make_pair(stats.PresentCount, stats));
+    if (!dxgi.tearing_support || !dxgi.force_disable_vsync) {
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
    if (!dxgi.tearing_support || dxgi.vsync_enabled) {
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 20]
```diff
@@ -422,150 +425,158 @@ static uint64_t qpc_to_100ns(uint64_t qpc) {
...
 
     return true;
 }
 
 static void gfx_dxgi_swap_buffers_begin(void) {
     // dxgi.length_in_vsync_frames = 1;
+
     LARGE_INTEGER t;
-    if (dxgi.use_timer) {
+    if (dxgi.force_disable_vsync) {
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
    if (!dxgi.vsync_enabled) {
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 21]
```diff
@@ -593,38 +623,40 @@ static void gfx_dxgi_swap_buffers_end(void) {
...
     QueryPerformanceCounter(&t1);
 
-    if (dxgi.applied_maximum_frame_latency > dxgi.maximum_frame_latency) {
-        // There seems to be a bug that if latency is decreased, there is no effect of that operation, so recreate swap
-        // chain
-        if (dxgi.waitable_object != nullptr) {
-            if (!dxgi.dropped_frame) {
-                // Wait the last time on this swap chain
-                WaitForSingleObject(dxgi.waitable_object, INFINITE);
+    if (!dxgi.tearing_support || !dxgi.force_disable_vsync) {
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
    if (!dxgi.tearing_support || dxgi.vsync_enabled) {
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 22]
```diff
@@ -709,6 +749,9 @@ void gfx_dxgi_create_swap_chain(IUnknown* device, std::function<void()>&& before
         dxgi.dxgi1_4 ? DXGI_SWAP_EFFECT_FLIP_DISCARD : // Introduced in DXGI 1.4 and Windows 10
             DXGI_SWAP_EFFECT_FLIP_SEQUENTIAL; // Apparently flip sequential was also backported to Win 7 Platform Update
     swap_chain_desc.Flags = dxgi_13 ? DXGI_SWAP_CHAIN_FLAG_FRAME_LATENCY_WAITABLE_OBJECT : 0;
+    if (dxgi.tearing_support && !dxgi.force_disable_vsync) {
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
    if (dxgi.tearing_support && dxgi.vsync_enabled) {
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 23]
```diff
@@ -761,6 +804,18 @@ const char* gfx_dxgi_get_key_name(int scancode) {
     return text;
 }
 
+bool gfx_dxgi_can_disable_vsync() {
+    return dxgi.tearing_support;
+}
+
+void gfx_dxgi_set_vsync(bool vsync) {
+    dxgi.force_disable_vsync = !vsync;
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
    dxgi.vsync_enabled = vsync;
```

File: src/graphic/Fast3D/gfx_dxgi.cpp [hunk 24]
```diff
@@ -761,6 +804,18 @@ const char* gfx_dxgi_get_key_name(int scancode) {
...
+}
+
+bool gfx_dxgi_get_vsync() {
+    return !dxgi.force_disable_vsync;
```
Comment (briaguya0) [inline src/graphic/Fast3D/gfx_dxgi.cpp]:
```suggestion
    return dxgi.vsync_enabled;
```

**Review by briaguya0 — COMMENTED** [review]
rename vsync stuff

**Comment by Kenix3** [comment]
superseded by https://github.com/Kenix3/libultraship/pull/171

**Pull request closed without merging**